Corrupted shard

After restarting Elasticsearch 1.5.2 (just a `sudo service elasticsearch restart`) on my single node server with about 600GB data one shard cannot get recovered and the cluster stays in the red status. I haven't noticed any problems before the restart. I'm using the indices without replicas.

I am using a GCE instance with persistent disks and there is enough free space for Elasticsearch there (>170GB). 

Error log:

    [2015-05-20 07:48:29,774][WARN ][indices.cluster          ] [test1] [[gardenhose-2015-week20][2]] marking and sending shard failed due to [failed recovery]
    org.elasticsearch.index.gateway.IndexShardGatewayRecoveryException: [gardenhose-2015-week20][2] failed recovery
            at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:162)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
            at java.lang.Thread.run(Thread.java:745)
    Caused by: org.elasticsearch.index.engine.EngineCreationFailureException: [gardenhose-2015-week20][2] failed to upgrade 3x segments
            at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:121)
            at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:32)
            at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:1262)
            at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:1257)
            at org.elasticsearch.index.shard.IndexShard.prepareForTranslogRecovery(IndexShard.java:784)
            at org.elasticsearch.index.gateway.local.LocalIndexShardGateway.recover(LocalIndexShardGateway.java:226)
            at org.elasticsearch.index.gateway.IndexShardGatewayService$1.run(IndexShardGatewayService.java:112)
            ... 3 more
    Caused by: java.io.EOFException: read past EOF: NIOFSIndexInput(path="/mnt/elasticsearch-test1/test/nodes/0/indices/gardenhose-2015-week20/2/index/segments_c9")
            at org.apache.lucene.store.BufferedIndexInput.refill(BufferedIndexInput.java:336)
            at org.apache.lucene.store.BufferedIndexInput.readByte(BufferedIndexInput.java:54)
            at org.apache.lucene.store.DataInput.readInt(DataInput.java:98)
            at org.apache.lucene.store.BufferedIndexInput.readInt(BufferedIndexInput.java:183)
            at org.elasticsearch.common.lucene.Lucene.indexNeeds3xUpgrading(Lucene.java:767)
            at org.elasticsearch.common.lucene.Lucene.upgradeLucene3xSegmentsMetadata(Lucene.java:778)
            at org.elasticsearch.index.engine.InternalEngine.upgrade3xSegments(InternalEngine.java:1084)
            at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:119)
            ... 9 more

--- a/src/main/java/org/elasticsearch/common/lucene/Lucene.java
+++ b/src/main/java/org/elasticsearch/common/lucene/Lucene.java
@@ -789,10 +789,10 @@ public class Lucene {
      * Returns <code>true</code> iff the store contains an index that contains segments that were
      * not upgraded to the lucene 4.x format.
      */
-    static boolean indexNeeds3xUpgrading(Directory directory) throws IOException {
-        final String si =  SegmentInfos.getLastCommitSegmentsFileName(directory);
-        if (si != null) {
-            try (IndexInput input = directory.openInput(si, IOContext.READONCE)) {
+    public static boolean indexNeeds3xUpgrading(Directory directory) throws IOException {
+        final String segmentsFile = SegmentInfos.getLastCommitSegmentsFileName(directory);
+        if (segmentsFile != null) {
+            try (IndexInput input = directory.openInput(segmentsFile, IOContext.READONCE)) {
                 return input.readInt() != CodecUtil.CODEC_MAGIC; // check if it's a 4.x commit point
             }
         }
@@ -801,20 +801,18 @@ public class Lucene {
 
     /**
      * Upgrades the segments metadata of the index to match a lucene 4.x index format. In particular it ensures that each
-     * segment has a .si file even if it was written with lucene 3.x
+     * segment has a .si file even if it was written with lucene 3.x.  Only call this if {@link #indexNeeds3xUpgrading}
+     * returned true.
      */
-    public static boolean upgradeLucene3xSegmentsMetadata(Directory directory) throws IOException {
-        if (indexNeeds3xUpgrading(directory)) {
-            try (final IndexWriter iw = new IndexWriter(directory, new IndexWriterConfig(Version.LATEST, Lucene.STANDARD_ANALYZER)
-                    .setMergePolicy(NoMergePolicy.INSTANCE)
-                    .setOpenMode(IndexWriterConfig.OpenMode.APPEND))) {
-                Map<String, String> commitData = iw.getCommitData(); // this is a trick to make IW to actually do a commit - we have to preserve the last committed data as well
-                // for ES to get the translog ID back
-                iw.setCommitData(commitData);
-                iw.commit();
-            }
-            return true;
+    public static void upgradeLucene3xSegmentsMetadata(Directory directory) throws IOException {
+        try (final IndexWriter iw = new IndexWriter(directory, new IndexWriterConfig(Version.LATEST, Lucene.STANDARD_ANALYZER)
+                                                    .setMergePolicy(NoMergePolicy.INSTANCE)
+                                                    .setOpenMode(IndexWriterConfig.OpenMode.APPEND))) {
+            Map<String, String> commitData = iw.getCommitData();
+            // this is a trick to make IW to actually do a commit - we have to preserve the last committed data as well
+            // for ES to get the translog ID back
+            iw.setCommitData(commitData);
+            iw.commit();
         }
-        return false;
     }
 }

--- a/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
+++ b/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
@@ -119,19 +119,12 @@ public class InternalEngine extends Engine {
         SearcherManager manager = null;
         boolean success = false;
         try {
-            try {
-                boolean autoUpgrade = true;
-                // If the index was created on 0.20.7 (Lucene 3.x) or earlier,
-                // it needs to be upgraded
-                autoUpgrade = Version.indexCreated(engineConfig.getIndexSettings()).onOrBefore(Version.V_0_20_7);
-                if (autoUpgrade) {
-                    logger.debug("[{}] checking for 3x segments to upgrade", shardId);
-                    upgrade3xSegments(store);
-                } else {
-                    logger.debug("[{}] skipping check for 3x segments", shardId);
-                }
-            } catch (IOException ex) {
-                throw new EngineCreationFailureException(shardId, "failed to upgrade 3x segments", ex);
+            // If the index was created on 0.20.7 (Lucene 3.x) or earlier, its commit point (segments_N file) needs to be upgraded:
+            if (Version.indexCreated(engineConfig.getIndexSettings()).onOrBefore(Version.V_0_20_7)) {
+                logger.debug("[{}] checking for 3x segments to upgrade", shardId);
+                maybeUpgrade3xSegments(store);
+            } else {
+                logger.debug("[{}] skipping check for 3x segments", shardId);
             }
             this.onGoingRecoveries = new FlushingRecoveryCounter(this, store, logger);
             this.lastDeleteVersionPruneTimeMSec = engineConfig.getThreadPool().estimatedTimeInMillis();
@@ -1139,10 +1132,25 @@ public class InternalEngine extends Engine {
         }
     }
 
-    protected void upgrade3xSegments(Store store) throws IOException {
+    protected void maybeUpgrade3xSegments(Store store) throws EngineException {
         store.incRef();
         try {
-            if (Lucene.upgradeLucene3xSegmentsMetadata(store.directory())) {
+            boolean doUpgrade;
+            try {
+                doUpgrade = Lucene.indexNeeds3xUpgrading(store.directory());
+            } catch (IOException ex) {
+                // This can happen if commit was truncated (e.g. due to prior disk full), and this case requires user intervention (remove the broken
+                // commit file so Lucene falls back to a previous good one, and also clear ES's corrupted_XXX marker file), and the shard
+                // should be OK:
+                throw new EngineCreationFailureException(shardId, "failed to read commit", ex);
+            }
+        
+            if (doUpgrade) {
+                try {
+                    Lucene.upgradeLucene3xSegmentsMetadata(store.directory());
+                } catch (IOException ex) {
+                    throw new EngineCreationFailureException(shardId, "failed to upgrade 3.x segments_N commit point", ex);
+                }
                 logger.debug("upgraded current 3.x segments file on startup");
             } else {
                 logger.debug("segments file is already after 3.x; not upgrading");

--- a/src/test/java/org/elasticsearch/common/lucene/LuceneTest.java
+++ b/src/test/java/org/elasticsearch/common/lucene/LuceneTest.java
@@ -330,14 +330,7 @@ public class LuceneTest extends ElasticsearchLuceneTestCase {
                         assertTrue(Lucene.indexNeeds3xUpgrading(dir));
                     }
 
-                    for (int i = 0; i < 2; i++) {
-                        boolean upgraded = Lucene.upgradeLucene3xSegmentsMetadata(dir);
-                        if (i == 0) {
-                            assertTrue(upgraded);
-                        } else {
-                            assertFalse(upgraded);
-                        }
-                    }
+                    Lucene.upgradeLucene3xSegmentsMetadata(dir);
 
                     for (int i = 0; i < 2; i++) {
                         assertFalse(Lucene.indexNeeds3xUpgrading(dir));

